Add documentation for Component SLIs feature (#37767)
kubernetes/website: content/en/docs/reference/instrumentation/slis.md

---
reviewers:
- logicalhan
title: Kubernetes Component SLI Metrics
linkTitle: Service Level Indicator Metrics
content_type: reference
weight: 20
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.26" state="alpha" >}}

As an alpha feature, Kubernetes lets you configure Service Level Indicator (SLI) metrics 
for each Kubernetes component binary. This metric endpoint is exposed on the serving 

> HTTPS port of each component, at the path `/metrics/slis`. You must enable to the 

Reviewer: nit
```suggestion
HTTPS port of each component, at the path `/metrics/slis`. You must enable the 
```

---
reviewers:
- logicalhan

> title: Kubernetes Component SLIs

Reviewer 2: (nit)
```suggestion
title: Kubernetes Component SLI Metrics
linkTitle: Service Level Indicator Metrics
```

content_type: reference
weight: 20
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.26" state="alpha" >}}

> The Component SLIs feature allows enabling of a new metrics endpoint for each Kubernetes

Reviewer 2: (nit)
```suggestion
As an alpha feature, Kubernetes lets you configure Service Level Indicator (SLI) metrics for each
Kubernetes component
```

> binary. This metric endpoint is exposed on the standard port of each binary at the endpoint

Reviewer 2: (nit)
```suggestion
binary. This metric endpoint is exposed on the serving HTTPS port of each component,
at the path
```

> `metrics/slis`. 

Reviewer 2: Is this right?
```suggestion
`/metrics/slis`. 
```

If not, readers may assume that the path is `/metrics/slis`

> 

Reviewer 2: ```suggestion
You must enable to the `ComponentSLIs` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
for every component from which you want to scrape SLI metrics.

```

<!-- body -->

## SLI Metrics

> We expose two metrics; a gauge (which represents the current state of the healthcheck) and a
> counter (which records the cumulative counts that we've observed for each healthcheck state).

Reviewer 2: (nit)

We [avoid using we](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we)

Is this right?
```suggestion
With SLI metrics enabled, each Kubernetes component exposes two metrics,
labelled per healthcheck:

- a gauge (which represents the current state of the healthcheck)
- a counter (which records the cumulative counts observed for each healthcheck state)
```

> 

Reviewer 2: (nit)
```suggestion

You can use the metric information to calculate per-component availability statistics.
For example, the API server checks the health of etcd. You can work out and report how
available or unavailable etcd has been - as reported by its client, the API server.

```

```golang
var (
   // healthcheck is a Prometheus Gauge metrics used for recording the results of a k8s healthcheck.
   healthcheck = k8smetrics.NewGaugeVec(
      &k8smetrics.GaugeOpts{
         Namespace:      "kubernetes",
         Name:           "healthcheck",
         Help:           "This metric records the result of a single healthcheck.",
         StabilityLevel: k8smetrics.ALPHA,
      },
      []string{"name", "type"},
   )

   // healthchecksTotal is a Prometheus Counter metrics used for counting the results of a k8s healthcheck.
   healthchecksTotal = k8smetrics.NewCounterVec(
      &k8smetrics.CounterOpts{
         Namespace:      "kubernetes",
         Name:           "healthchecks_total",
         Help:           "This metric records the results of all healthcheck.",
         StabilityLevel: k8smetrics.ALPHA,
      },
      []string{"name", "type", "status"},
   )

> )

Reviewer 2: It's OK to omit the source code; we don't expect readers to know how to code in Golang. We could hyperlink to the implementation if that's important.

Please either omit this or make it clear why we should include it.

Author: I actually do expect people who interact with prometheus metrics to be able to read the metric definitions. It's mostly declarative and easier to parse than most random code.

Author: removed

```

The gauge data looks like this:

> ```shell

Reviewer 2: This is actually OpenMetrics (or something close), not shell. It's OK to omit a highlighting hint. Please don't set one that's wrong though.

Author: There isn't syntax highlight for prometheus format, but I found shell colorizes very close to correctly.

# HELP kubernetes_healthcheck [ALPHA] This metric records the result of a single healthcheck.
# TYPE kubernetes_healthcheck gauge
kubernetes_healthcheck{name="autoregister-completion",type="healthz"} 1
kubernetes_healthcheck{name="autoregister-completion",type="readyz"} 1
kubernetes_healthcheck{name="etcd",type="healthz"} 1
kubernetes_healthcheck{name="etcd",type="readyz"} 1
kubernetes_healthcheck{name="etcd-readiness",type="readyz"} 1
kubernetes_healthcheck{name="informer-sync",type="readyz"} 1
kubernetes_healthcheck{name="log",type="healthz"} 1
kubernetes_healthcheck{name="log",type="readyz"} 1
kubernetes_healthcheck{name="ping",type="healthz"} 1
kubernetes_healthcheck{name="ping",type="readyz"} 1
```

While the counter data looks like this:

> ```shell

Reviewer 2: (nit)
This is actually OpenMetrics (or something close), not shell. It's OK to omit a highlighting hint. Please don't set one that's wrong though.

Author: It colorizes pretty correctly under this syntax though. It looks nicer than if I omit here.

Author: removed

kubernetes_healthchecks_total{name="autoregister-completion",status="success",type="healthz"} 15
kubernetes_healthchecks_total{name="autoregister-completion",status="success",type="readyz"} 14
kubernetes_healthchecks_total{name="etcd",status="success",type="healthz"} 15
kubernetes_healthchecks_total{name="etcd",status="success",type="readyz"} 15
kubernetes_healthchecks_total{name="etcd-readiness",status="success",type="readyz"} 15
kubernetes_healthchecks_total{name="informer-sync",status="error",type="readyz"} 1
kubernetes_healthchecks_total{name="informer-sync",status="success",type="readyz"} 14
kubernetes_healthchecks_total{name="log",status="success",type="healthz"} 15
kubernetes_healthchecks_total{name="log",status="success",type="readyz"} 15
kubernetes_healthchecks_total{name="ping",status="success",type="healthz"} 15
kubernetes_healthchecks_total{name="ping",status="success",type="readyz"} 15
```

## Using this data

The component SLIs metrics endpoint is intended to be scraped at a high frequency. Scraping
at a high frequency means that you end up with greater granularity of the gauge's signal, which

> can be then used to calculate SLOs. The `metrics/slis` endpoint provides the raw data necessary

Reviewer 2: ```suggestion
can be then used to calculate SLOs. The `/metrics/slis` endpoint provides the raw data necessary
```

> to calculate an availability SLO for the respective Kubernetes component. 

Reviewer 2: (nit)
What's a high frequency? Are you thinking tens of scrapes per minute, or tens of scrapes per second?
